Scripts/RemediationScripts/Remediate-EnableTransparentDataEncryption.ps1

```diff
@@ -0,0 +1,653 @@
+
+        [Switch]
+        [Parameter(ParameterSetName = "DryRun", HelpMessage="Specifies validation of prerequisites for the command")]
+        [Parameter(ParameterSetName = "WetRun", HelpMessage="Specifies validation of prerequisites for the command")]
+        $PerformPreReqCheck,
+
+        [Switch]
+        [Parameter(ParameterSetName = "DryRun", Mandatory = $true, HelpMessage="Specifies a dry run of the actual remediation")]
+        $DryRun,
+
+        [String]
+        [Parameter(ParameterSetName = "WetRun", HelpMessage="Specifies the path to the file to be used as input for the remediation")]
+        $FilePath
+    )
+
+    Write-Host $([Constants]::DoubleDashLine)
+    Write-Host "[Step 1 of 4] Preparing to enable Transparent data encryption (TDE) for SQL Servers in Subscription: $($SubscriptionId)"
+
+    if ($PerformPreReqCheck)
+    {
+        try
+        {
+            Write-Host "Setting up prerequisites..."
+            Setup-Prerequisites
+        }
+        catch
+        {
+            Write-Host "Error occurred while setting up prerequisites. Error: $($_)" -ForegroundColor $([Constants]::MessageType.Error)
+            break
+        }
+    }
+
+    # Connect to Azure account
+    $context = Get-AzContext
+
+    if ([String]::IsNullOrWhiteSpace($context))
+    {
+        Write-Host "Connecting to Azure account..."
+        Connect-AzAccount -Subscription $SubscriptionId -ErrorAction Stop | Out-Null
+        Write-Host "Connected to Azure account." -ForegroundColor $([Constants]::MessageType.Update)
+    }
+
+    # Setting up context for the current Subscription.
+    $context = Set-AzContext -SubscriptionId $SubscriptionId -ErrorAction Stop
+    
+    Write-Host $([Constants]::SingleDashLine)
+    Write-Host "Subscription Name: $($context.Subscription.Name)"
+    Write-Host "Subscription ID: $($context.Subscription.SubscriptionId)"
+    Write-Host "Account Name: $($context.Account.Id)"
+    Write-Host "Account Type: $($context.Account.Type)"
+    Write-Host $([Constants]::SingleDashLine)
+
+    Write-Host "Checking if $($context.Account.Id) is allowed to run this script..."
+
+    # Checking if the current account type is "User"
+    if ($context.Account.Type -ne "User")
+    {
+        Write-Host "WARNING: This script can only be run by `User` Account Type. Account Type of $($context.Account.Id) is: $($context.Account.Type)" -ForegroundColor $([Constants]::MessageType.Warning)
+        break
+    }
+
+    Write-Host "*** To enable Transparent data encryption (TDE) for SQL Servers in a Subscription, Contributor and higher privileges on the SQL Servers are required. ***" -ForegroundColor $([Constants]::MessageType.Info)
+    Write-Host $([Constants]::DoubleDashLine)
+    Write-Host "[Step 2 of 4] Preparing to fetch all SQL Servers..."
+
+    $sqlServersResources = @()
+    $databaseList = @()
+
+    # Includes SQL Servers where Transparent Data Encryption (TDE) is enabled.
+    $sqlDatabaseWithTDEEnabled = @()
```

Comment:
Would it be more appropriate to maintain just one record at the SQL Server level and not one record for every SQL Databases within them? For a simple reason, if anyone wants to know just the total non-compliant SQL Servers count in a Subscription, rather than the total non-compliant SQL Databases count across all SQL Servers.

If still using the current approach, `sqlDatabaseWithTDEEnabled` to be replaced with `sqlDatabasesWithTDEEnabled`, `sqlDatabaseWithTDEDisabled` with `sqlDatabasesWithTDEDisabled`.

Also, `sqlServersSkipped` is to be renamed to maintain uniformity with the above.